### PR TITLE
exp/client/horizon: operations

### PR DESCRIPTION
**Pull request by poliha**
This PR adds support for retrieving operations from horizon. It supports operations for accounts, ledgers, transactions and single operation id.

Horizon can return dynamic operation types per request, so we needed a way to make it easy for the user to access the responses without doing type assertions. 
A custom `UnmarshalJSON` method for `OperationsPage` was implemented to help with this, see `protocols/horizon/operations/main.go`.  This will be extended in a separate PR for `protocols/horizon`
The same concept is used when returning responses for a single operation request in `clients.go`

Some feedback on this will be aprreciated.

### CODE REVIEW

File: exp/clients/horizon/client.go [hunk 1]
```diff
@@ -136,3 +139,74 @@ func (c *Client) Offers(request OfferRequest) (offers hProtocol.OffersPage, err
...
+// Operations returns stellar operations(https://www.stellar.org/developers/horizon/reference/resources/operation.html)
+// It can be used to return operations for an account, a ledger, a transaction and all operations on the network.
+func (c *Client) Operations(request OperationRequest) (ops operations.OperationsPage, err error) {
+	err = c.sendRequest(request, &ops)
+	return
+}
+
+// OperationDetail returns a single stellar operations(https://www.stellar.org/developers/horizon/reference/resources/operation.html)
+// for a given operaion id
+func (c *Client) OperationDetail(id uint) (ops operations.Operation, err error) {
```
Comment (bartekn) [inline exp/clients/horizon/client.go]: I think we should make `id` a `string`. It's currently numbers only (ID is generated by `toid` package) but there's no guarantee that it will stay like this. Also, `Base.ID` is `string`.

BTW. ID currently returned by `toid` is 64-bit long so should be `uint64`, from [docuentation](https://golang.org/pkg/builtin/#uint):
> uint is an unsigned integer type that is at least 32 bits in size. It is a distinct type, however, and not an alias for, say, uint32.

So it can be `uint32` on some machines but `uint64` on others.

File: exp/clients/horizon/client.go [hunk 2]
```diff
@@ -136,3 +139,74 @@ func (c *Client) Offers(request OfferRequest) (offers hProtocol.OffersPage, err
...
+			return
+		}
+		ops = op
+	case operations.TypeNames[xdr.OperationTypeChangeTrust]:
+		var op operations.ChangeTrust
+		if err = json.Unmarshal(dataString, &op); err != nil {
+			return
+		}
+		ops = op
+	default:
```
Comment (bartekn) [inline exp/clients/horizon/client.go]: Default should return error. Add all operation types.

File: protocols/horizon/operations/main.go [hunk 3]
```diff
@@ -174,3 +176,120 @@ type AccountMerge struct {
...
+				return err
+			}
+			ops.Embedded.Records = append(ops.Embedded.Records, op)
+		case TypeNames[xdr.OperationTypePayment]:
+			var op Payment
+			if err := json.Unmarshal(dataString, &op); err != nil {
+				return err
+			}
+			ops.Embedded.Records = append(ops.Embedded.Records, op)
+		case TypeNames[xdr.OperationTypeManageOffer]:
```
Comment (bartekn) [inline protocols/horizon/operations/main.go]: Can you move this code to a function? It is exactly the same in `OperationDetail`.

File: exp/clients/horizon/effect_request.go [hunk 4]
```diff
@@ -112,7 +112,7 @@ func stream(
 		}
 
 		// Expected statusCode are 200-299
-		if resp.StatusCode >= 200 && resp.StatusCode < 300 {
+		if resp.StatusCode < 200 && resp.StatusCode >= 300 {
```
Comment (bartekn) [inline exp/clients/horizon/effect_request.go]: This will never be true. Should be `||` instead of `&&` or negate original line.

File: exp/clients/horizon/internal.go [hunk 5]
```diff
@@ -34,9 +34,24 @@ func decodeResponse(resp *http.Response, object interface{}) (err error) {
...
+			if param > 0 {
+				counter++
+			}
+		case uint:
+			if param > 0 {
+				counter++
+			}
+		case bool:
 			counter++
+		default:
```
Comment (bartekn) [inline exp/clients/horizon/internal.go]: I think it should `panic` in this case.

File: exp/clients/horizon/main.go [hunk 6]
```diff
@@ -154,3 +160,15 @@ type OfferRequest struct {
...
+// OperationRequest struct contains data for getting operation details from an horizon servers
+type OperationRequest struct {
+	ForAccount     string
+	ForLedger      int
+	ForTransaction string
+	forOperationId uint
+	Order          Order
+	Cursor         Cursor
+	Limit          Limit
+	IncludeFailed  IncludeFailed
```
Comment (bartekn) [inline exp/clients/horizon/main.go]: EDIT: OK, I think it's used in `addQueryParams`. Can we change the types to private types and use them just internally?

---
I think I missed that in the earlier PRs but I don't see any clear advantage of using extra types for `Cursor`, `Limit` and `IncludeFailed`. We had a call like this in the previous client because we were using variadic to pass params like:
```go
client.Transactions(Cursor("1234"), Limit(10))
```
In the current code we don't need this because each param is a separate field on request struct.

This is not the case for `Order` as there are only 2 possible values and it's good to pass const like `OrderAsc` or `OrderDesc` as typos will be caught by a compiler.

File: exp/clients/horizon/main_test.go [hunk 7]
```diff
@@ -151,6 +152,55 @@ func ExampleClient_Offers() {
 	fmt.Print(offers)
 }
 
+func ExampleClient_Operations() {
+
```
Comment (bartekn) [inline exp/clients/horizon/main_test.go]: Let's leave it like this for now but this function will be run during tests so 1) it shouldn't make any network connections (maybe use a mock?), 2) it's actual test as you can compare actual and expected output using `// Output:` comment: https://blog.golang.org/examples.

File: protocols/horizon/operations/main.go [hunk 8]
```diff
@@ -174,3 +176,120 @@ type AccountMerge struct {
 type Inflation struct {
 	Base
 }
+
+// Operation interface contains methods implemented by the operation types
+type Operation interface {
+	PagingToken() string
+	GetType() string
```
Comment (bartekn) [inline protocols/horizon/operations/main.go]: I think we shouldn't have methods below this one. First, the looks like static methods and not per instance methods. Second, users should really use type assertions or type switches to get actual object like:
```go
createAccount, ok := op.(CreateAccount)
if ok {
   fmt.Println(createAccount.StartingBalance)
}
```

Comment (poliha) [reply protocols/horizon/operations/main.go]: Okay that's fair enough, besides to use this correctly, the user still has to know the type so they might as well do the type assertion.  I might add more methods for getting common operation parameters though; e.g for successful transaction

File: exp/clients/horizon/main_test.go (hunk 7 above)
Comment (bartekn) [reply exp/clients/horizon/main_test.go]: OK, cool!

File: protocols/horizon/operations/main.go (hunk 8 above)
Comment (bartekn) [reply protocols/horizon/operations/main.go]: > besides to use this correctly, the user still has to know the type so they might as well do the type assertion

Yeah, they can use type switch or use `GetType()` method.

**Comment by poliha** [comment]
Okay @bartekn, made the requested changes. If you can, please have one more look through as the changes where a bit significant before I merge.

File: exp/clients/horizon/client.go [hunk 9]
```diff
@@ -136,3 +138,51 @@ func (c *Client) Offers(request OfferRequest) (offers hProtocol.OffersPage, err
...
+// Operations returns stellar operations (https://www.stellar.org/developers/horizon/reference/resources/operation.html)
+// It can be used to return operations for an account, a ledger, a transaction and all operations on the network.
+func (c *Client) Operations(request OperationRequest) (ops operations.OperationsPage, err error) {
+	err = c.sendRequest(request, &ops)
+	return
+}
+
+// OperationDetail returns a single stellar operations (https://www.stellar.org/developers/horizon/reference/resources/operation.html)
+// for a given operation id
+func (c *Client) OperationDetail(id string) (ops operations.Operation, err error) {
```
Comment (howardtw) [inline exp/clients/horizon/client.go]: For long functions, we should avoid using named return arguments.
```
Naked return statements should be used only in short functions, as with the example shown here. They can harm readability in longer functions.
```

File: exp/clients/horizon/client.go [hunk 10]
```diff
@@ -136,3 +138,51 @@ func (c *Client) Offers(request OfferRequest) (offers hProtocol.OffersPage, err
...
+	if id == "" {
+		err = errors.New("Invalid operation id provided")
+	}
+
+	if err != nil {
```
Comment (howardtw) [inline exp/clients/horizon/client.go]: Is this condition ever going to be true?

File: exp/clients/horizon/client.go [hunk 11]
```diff
@@ -136,3 +138,51 @@ func (c *Client) Offers(request OfferRequest) (offers hProtocol.OffersPage, err
...
+	}
+
+	request := OperationRequest{forOperationId: id}
+
+	var record interface{}
+
+	err = c.sendRequest(request, &record)
+
+	if err != nil {
+		return
```
Comment (howardtw) [inline exp/clients/horizon/client.go]: We should use `Wrap` and include a message. This applies to all the error cases return in this function :)

File: exp/clients/horizon/internal.go [hunk 12]
```diff
@@ -34,9 +34,25 @@ func decodeResponse(resp *http.Response, object interface{}) (err error) {
 func countParams(params ...interface{}) int {
 	counter := 0
 	for _, param := range params {
-		if param != "" {
+		switch param := param.(type) {
+		case string:
+			if param != "" {
+				counter++
+			}
+		case int:
```
Comment (howardtw) [inline exp/clients/horizon/internal.go]: Is it possible to combine `int` and `uint`?

File: exp/clients/horizon/operation_request.go [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+	"fmt"
+	"net/url"
+
+	"github.com/stellar/go/support/errors"
+)
+
+// BuildUrl creates the endpoint to be queried based on the data in the OperationRequest struct.
+// If no data is set, it defaults to the build the URL for all operations
+func (op OperationRequest) BuildUrl() (endpoint string, err error) {
+
```
Comment (howardtw) [inline exp/clients/horizon/operation_request.go]: extra line

File: exp/clients/horizon/client.go (hunk 9 above)
Comment (poliha) [reply exp/clients/horizon/client.go]: Will take note and review in a separate PR as there might be more instances of this.

File: exp/clients/horizon/client.go (hunk 10 above)
Comment (poliha) [reply exp/clients/horizon/client.go]: Yes if an empty string is passed as the `operaionId`

File: exp/clients/horizon/internal.go (hunk 12 above)
Comment (poliha) [reply exp/clients/horizon/internal.go]: Tried this earlier. the param check will have a syntax error `operator > not defined on interface`

File: exp/clients/horizon/client.go (hunk 10 above)
Comment (howardtw) [reply exp/clients/horizon/client.go]: I mean why don’t we return at line 153?

Comment (poliha) [reply exp/clients/horizon/client.go]: Oh I see what you mean now. Just conforming with the style I have seen in other packages. Also its double checking that `errors.New` is working as expected.

File: exp/clients/horizon/main.go [hunk 14]
```diff
@@ -118,24 +124,24 @@ type EffectRequest struct {
 	ForOperation   string
 	ForTransaction string
 	Order          Order
-	Cursor         Cursor
-	Limit          Limit
+	Cursor         cursor
+	Limit          limit
```
Comment (bartekn) [inline exp/clients/horizon/main.go]: Sorry, I think I wasn't clear. The exported structs have standard types like `int`, `string` and exported types (`Order`). When user pass string `Cursor` you then convert it internally to `cursor`.

I think that currently users won't be able to create this object because types are not exported.

File: exp/clients/horizon/main.go [hunk 15]
```diff
@@ -151,6 +157,18 @@ type feeStatsRequest struct {
...
+
+// OperationRequest struct contains data for getting operation details from an horizon servers
+type OperationRequest struct {
+	ForAccount     string
+	ForLedger      int
+	ForTransaction string
+	forOperationId string
+	Order          Order
+	Cursor         cursor
+	Limit          limit
```
Comment (bartekn) [inline exp/clients/horizon/main.go]: I think I added a comment about it in the past: can we move common fields like `Order`, `Cursor`, etc. to another struct that will be later embedded in `*Request` structs?

File: protocols/horizon/operations/main.go [hunk 16]
```diff
@@ -174,3 +176,153 @@ type AccountMerge struct {
...
+
+		op, err := UnmarshalOperation(b.Type, dataString)
+		if err != nil {
+			return err
+		}
+
+		ops.Embedded.Records = append(ops.Embedded.Records, op)
+	}
+
+	return nil
```
Comment (bartekn) [inline protocols/horizon/operations/main.go]: It seems that `Links` are not copied to `ops`. Can you add tests checking links?

Comment (poliha) [reply protocols/horizon/operations/main.go]: okay will check it out

**Comment by poliha** [comment]
Okay @bartekn @howardtw made some more changes.... Please have a look when you can

File: exp/clients/horizon/main.go [hunk 17]
```diff
@@ -117,25 +130,19 @@ type EffectRequest struct {
 	ForLedger      string
 	ForOperation   string
 	ForTransaction string
-	Order          Order
-	Cursor         Cursor
-	Limit          Limit
+	RequestParams
```
Comment (bartekn) [inline exp/clients/horizon/main.go]: Really sorry, @poliha. Can you revert this change? I was hoping that you can use the fields of embedded structs directly but it seems that something like this is no possible:
```go
req := LedgerRequest{
   Limit: 100,
}
```

File: protocols/horizon/operations/main.go [hunk 18]
```diff
@@ -174,3 +176,154 @@ type AccountMerge struct {
...
+			return err
+		}
+
+		op, err := UnmarshalOperation(b.Type, dataString)
+		if err != nil {
+			return err
+		}
+
+		ops.Embedded.Records = append(ops.Embedded.Records, op)
+		ops.Links = opsPage.Links
```
Comment (bartekn) [inline protocols/horizon/operations/main.go]: You can move it outside this loop. There's a single `_links` field in page response.

Comment (poliha) [reply protocols/horizon/operations/main.go]: good catch .. thanks

File: exp/clients/horizon/main.go (hunk 17 above)
Comment (poliha) [reply exp/clients/horizon/main.go]: :) no problem.
Yea that's not possible with embedded structs.. You can do this though
```golang
req := LedgerRequest{}
req.Limit = 100
```

File: exp/clients/horizon/client.go (hunk 9 above)
Comment (howardtw) [reply exp/clients/horizon/client.go]: I would prefer to address it in this PR :)

File: exp/clients/horizon/client.go [hunk 19]
```diff
@@ -136,3 +138,55 @@ func (c *Client) Offers(request OfferRequest) (offers hProtocol.OffersPage, err
...
+func (c *Client) Operations(request OperationRequest) (ops operations.OperationsPage, err error) {
+	err = c.sendRequest(request, &ops)
+	return
+}
+
+// OperationDetail returns a single stellar operations (https://www.stellar.org/developers/horizon/reference/resources/operation.html)
+// for a given operation id
+func (c *Client) OperationDetail(id string) (ops operations.Operation, err error) {
+	if id == "" {
+		err = errors.New("Invalid operation id provided")
```
Comment (howardtw) [inline exp/clients/horizon/client.go]: we can do `return ops, errors.New(...)` and remove the following block

File: exp/clients/horizon/client.go [hunk 20]
```diff
@@ -136,3 +138,55 @@ func (c *Client) Offers(request OfferRequest) (offers hProtocol.OffersPage, err
...
+	if err != nil {
+		return ops, err
+	}
+
+	request := OperationRequest{forOperationId: id}
+
+	var record interface{}
+
+	err = c.sendRequest(request, &record)
+
```
Comment (howardtw) [inline exp/clients/horizon/client.go]: extra line

File: exp/clients/horizon/client.go [hunk 21]
```diff
@@ -136,3 +138,55 @@ func (c *Client) Offers(request OfferRequest) (offers hProtocol.OffersPage, err
...
+	if err != nil {
+		err = errors.Wrap(err, "Sending request to horizon")
```
Comment (howardtw) [inline exp/clients/horizon/client.go]: we can do `return ops, errors.Wrap(err, ...)`. This applies to all the returns :)

File: exp/clients/horizon/client.go [hunk 22]
```diff
@@ -136,3 +138,55 @@ func (c *Client) Offers(request OfferRequest) (offers hProtocol.OffersPage, err
...
+		err = errors.Wrap(err, "Marshaling json")
+		return ops, err
+	}
+	if err = json.Unmarshal(dataString, &baseRecord); err != nil {
+		err = errors.Wrap(err, "Unmarshaling json")
+		return ops, err
+	}
+
+	// unmarshal to the correct json struct based on operation type
+	op, err := operations.UnmarshalOperation(baseRecord.GetType(), dataString)
```
Comment (howardtw) [inline exp/clients/horizon/client.go]: Is it possible to do this?
```
ops, err := operations.UnmarshalOperation(baseRecord.GetType(), dataString)
return ops, errors.Wrap(err, "Unmarshaling to the correct operation type")
```

File: exp/clients/horizon/operation_request.go [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+import (
+	"fmt"
+	"net/url"
+
+	"github.com/stellar/go/support/errors"
+)
+
+// BuildUrl creates the endpoint to be queried based on the data in the OperationRequest struct.
+// If no data is set, it defaults to the build the URL for all operations
+func (op OperationRequest) BuildUrl() (endpoint string, err error) {
```
Comment (howardtw) [inline exp/clients/horizon/operation_request.go]: Can we address the named returned values here too?

File: exp/clients/horizon/operation_request.go [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+	nParams := countParams(op.ForAccount, op.ForLedger, op.forOperationId, op.ForTransaction)
+
+	if nParams > 1 {
+		err = errors.New("Invalid request. Too many parameters")
```
Comment (howardtw) [inline exp/clients/horizon/operation_request.go]: We should return directly here and remove the following block.

File: exp/clients/horizon/operation_request.go [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+	}
+
+	if err != nil {
+		return endpoint, err
+	}
+
+	endpoint = "operations"
+
+	if op.ForAccount != "" {
+		endpoint = fmt.Sprintf(
```
Comment (howardtw) [inline exp/clients/horizon/operation_request.go]: Maybe we don't have to use multiple lines here? `endpoint = fmt.Sprintf(`"accounts/%s/operations", op.ForAccount` reads perfectly fine to me

File: exp/clients/horizon/client.go (hunk 22 above)
Comment (poliha) [reply exp/clients/horizon/client.go]: Yep this should work

File: exp/clients/horizon/operation_request.go (hunk 23 above)
Comment (poliha) [reply exp/clients/horizon/operation_request.go]: I believe we already do `return endpoint, err`

File: exp/clients/horizon/operation_request.go (hunk 25 above)
Comment (poliha) [reply exp/clients/horizon/operation_request.go]: Sounds good

File: exp/clients/horizon/main.go [hunk 26]
```diff
@@ -118,24 +124,24 @@ type EffectRequest struct {
...
+	Cursor         string
+	Limit          uint
 }
 
 // AssetRequest struct contains data for getting asset details from an horizon server.
 type AssetRequest struct {
-	ForAssetCode   AssetCode
-	ForAssetIssuer AssetIssuer
+	ForAssetCode   assetCode
+	ForAssetIssuer assetIssuer
```
Comment (bartekn) [inline exp/clients/horizon/main.go]: These should be strings. You shouldn't export values that are of unexported type. I'm pretty sure users won't be able to create it.

File: exp/clients/horizon/main.go [hunk 27]
```diff
@@ -79,17 +83,19 @@ type ClientInterface interface {
...
 	FeeStats() (hProtocol.FeeStats, error)
 	Offers(request OfferRequest) (hProtocol.OffersPage, error)
+	Operations(request OperationRequest) (operations.OperationsPage, error)
+	OperationDetail(id string) (operations.Operation, error)
 }
 
 // DefaultTestNetClient is a default client to connect to test network
 var DefaultTestNetClient = &Client{
-	HorizonURL: "https://horizon-testnet.stellar.org",
+	HorizonURL: "https://horizon-testnet.stellar.org/",
```
Comment (bartekn) [inline exp/clients/horizon/main.go]: Are we prepared for situations like this? I.e. if I pass horizon server URL without trailing slash would it work?

File: exp/clients/horizon/main.go (hunk 26 above)
Comment (poliha) [reply exp/clients/horizon/main.go]: Oh right I missed that, when I reverted my last commit. Will fix.

File: exp/clients/horizon/main.go (hunk 27 above)
Comment (poliha) [reply exp/clients/horizon/main.go]: Currently it wouldn't work.. Plan to fix this in my another PR, similar to `fixURLOnce` in the old client
